src/xtgeo/io/_welldata/_blockedwell_io.py

```diff
@@ -0,0 +1,114 @@
+"""I/O of blocked well data"""
+
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+
+import numpy as np
+import numpy.typing as npt
+
+from xtgeo.common.log import null_logger
+from xtgeo.io._welldata._well_io import WellData
+
+logger = null_logger(__name__)
+
+
+@dataclass(frozen=True)
+class BlockedWellData(WellData):
+    """Immutable data container for blocked well data.
+
+    This dataclass extends WellData with grid cell indices, representing a well
+    that has been intersected with a 3D grid.
+
+    This class is derived from WellData, but a difference for BlockedWell is that
+    each survey point is associated with a grid cell location defined by
+    I, J, K indices.
+
+    Attributes:
+        name: Well name (inherited from WellData)
+        xpos: X-coordinate of the well header position (inherited)
+        ypos: Y-coordinate of the well header position (inherited)
+        zpos: Z-coordinate of the well header position (inherited)
+        survey_x: X-coordinates along wellbore trajectory (inherited)
+        survey_y: Y-coordinates along wellbore trajectory (inherited)
+        survey_z: Z-coordinates along wellbore trajectory (inherited)
+        logs: Tuple of WellLog objects (inherited)
+        i_index: NumPy array of I-indices (column) for grid cells
+        j_index: NumPy array of J-indices (row) for grid cells
+        k_index: NumPy array of K-indices (layer) for grid cells
+
+    Notes:
+        - Grid indices should use 1-based indexing (typical in reservoir modeling)
+        - Undefined/inactive cells can be represented with np.nan or a negative value
+        - All index arrays must have the same length as survey arrays
+        - Index arrays are typically integer-valued but stored as float to allow NaN
+    """
+
+    i_index: npt.NDArray[np.float64] = field(default=None)  # type: ignore[arg-type]
+    j_index: npt.NDArray[np.float64] = field(default=None)  # type: ignore[arg-type]
+    k_index: npt.NDArray[np.float64] = field(default=None)  # type: ignore[arg-type]
+
+    def __post_init__(self) -> None:
+        """Validate blocked well data consistency."""
+        # First call parent validation
+        super().__post_init__()
+
+        # Validate that index arrays are provided
+        if self.i_index is None or self.j_index is None or self.k_index is None:
+            raise ValueError(
+                "BlockedWellData requires i_index, j_index, and k_index arrays"
+            )
```

Comment:
The `__init__` will fail if any of these aren't provided, so I think this check is not necessary. It's only necessary if the fields are given a default (as they currently are)